src/passes/GlobalEffects.cpp

```diff
@@ -22,60 +22,184 @@
+      }
+    }
+
+    // Compute the transitive closure of the call graph, that is, fill out
+    // |callers| so that it contains the list of all callers - even through a
+    // chain - of each function.
+    while (!work.empty()) {
+      auto [caller, called] = work.pop();
+
+      // We must not already have an entry for this call (that would imply we
+      // are doing wasted work).
+      assert(!callers[called].count(caller));
+
+      // Apply the new call information.
+      callers[called].insert(caller);
+
+      // We just learned that |caller| calls |called|. It also calls
+      // transitively, which we need to propagate to all places unaware of that
+      // information yet.
+      //
+      //   caller => called => called by called
+      //
+      auto& calledInfo = analysis.map[module->getFunction(called)];
+      for (auto calledByCalled : calledInfo.calledFunctions) {
+        if (!callers[calledByCalled].count(caller)) {
+          work.push({caller, calledByCalled});
+        }
+      }
+    }
+
+    // Now that we have transitively propagated all static calls, apply that
+    // information. First, apply infinite recursion: if a function can call
+    // itself then it might recurse infinitely, which we consider an effect (a
+    // trap).
+    for (auto& [func, info] : analysis.map) {
+      if (callers[func->name].count(func->name)) {
+        if (info.effects) {
+          info.effects->trap = true;
+        }
+      }
+    }
+
+    // Next, apply function effects to their callers.
+    for (auto& [func, info] : analysis.map) {
+      auto& funcEffects = info.effects;
+
+      for (auto& caller : callers[func->name]) {
+        auto& callerEffects = analysis.map[module->getFunction(caller)].effects;
+        if (!callerEffects) {
+          // Nothing is known for the caller, which is already the worst case.
+          continue;
         }
-        funcEffectsMap->emplace(func->name, *possibleEffects);
+
+        if (!funcEffects) {
+          // Nothing is known for the called function, which means nothing is
+          // known for the caller either.
+          callerEffects.reset();
+          continue;
+        }
+
+        // Add func's effects to the caller.
+        callerEffects->mergeIn(*funcEffects);
+      }
+    }
+
+    // Generate the final data structure, starting from a blank slate where
+    // nothing is known.
+    auto& funcEffectsMap = getPassOptions().funcEffectsMap;
+    funcEffectsMap.reset();
+
+    for (auto& [func, info] : analysis.map) {
+      if (!info.effects) {
+        // Add no entry to funcEffectsMap, since nothing is known.
+        continue;
+      }
+
+      // Only allocate a new funcEffectsMap here when we actually have data for
+      // it (which might make later effect computation slightly faster, to
+      // quickly skip the funcEffectsMap code path).
+      if (!funcEffectsMap) {
+        funcEffectsMap = std::make_shared<FuncEffectsMap>();
```

Comment:
Why is this a shared pointer? I would have assumed we would use unique ownership and anything that needs to reference it would do so via a normal pointer or reference.

Reply:
I think the reason (predating this PR of course) is that we store this on a `PassOption`, and we do allow copying those around.

Reply:
Makes sense, thanks 👍